{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook contains code for creating a promoter library for promoter bashing.\n",
    "It details how to create libraries of <300 nucleotides that can be ordered, as well as creating the primers for cloning these into vectors."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# get transcription factors and motifs.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## find transcription factors and their imputed motifs from plantregdb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>Family</th>\n",
       "      <th>Matrix_id</th>\n",
       "      <th>Species</th>\n",
       "      <th>Method</th>\n",
       "      <th>Datasource</th>\n",
       "      <th>Datasource_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Sobic.001G002500</td>\n",
       "      <td>ERF</td>\n",
       "      <td>MP00374</td>\n",
       "      <td>Sorghum bicolor</td>\n",
       "      <td>DAP</td>\n",
       "      <td>PlantTFDB</td>\n",
       "      <td>transfer from AT3G23240(Arabidopsis thaliana)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Sobic.001G007200</td>\n",
       "      <td>HSF</td>\n",
       "      <td>MP00514</td>\n",
       "      <td>Sorghum bicolor</td>\n",
       "      <td>DAP</td>\n",
       "      <td>PlantTFDB</td>\n",
       "      <td>transfer from AT5G16820(Arabidopsis thaliana)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Sobic.001G026500</td>\n",
       "      <td>SBP</td>\n",
       "      <td>MP00633</td>\n",
       "      <td>Sorghum bicolor</td>\n",
       "      <td>PBM</td>\n",
       "      <td>PlantTFDB</td>\n",
       "      <td>transfer from PK06791.1(Cannabis sativa)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Sobic.001G034300</td>\n",
       "      <td>Dof</td>\n",
       "      <td>MP00295</td>\n",
       "      <td>Sorghum bicolor</td>\n",
       "      <td>DAP</td>\n",
       "      <td>PlantTFDB</td>\n",
       "      <td>transfer from AT2G37590(Arabidopsis thaliana)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Sobic.001G036800</td>\n",
       "      <td>AP2</td>\n",
       "      <td>MP00615</td>\n",
       "      <td>Sorghum bicolor</td>\n",
       "      <td>PBM</td>\n",
       "      <td>PlantTFDB</td>\n",
       "      <td>transfer from AT2G28550(Arabidopsis thaliana)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>255</th>\n",
       "      <td>Sobic.010G215700</td>\n",
       "      <td>SBP</td>\n",
       "      <td>MP00090</td>\n",
       "      <td>Sorghum bicolor</td>\n",
       "      <td>SELEX</td>\n",
       "      <td>PlantTFDB</td>\n",
       "      <td>transfer from AT1G02065(Arabidopsis thaliana)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>256</th>\n",
       "      <td>Sobic.010G236300</td>\n",
       "      <td>ARF</td>\n",
       "      <td>MP00461</td>\n",
       "      <td>Sorghum bicolor</td>\n",
       "      <td>DAP</td>\n",
       "      <td>PlantTFDB</td>\n",
       "      <td>transfer from AT4G30080(Arabidopsis thaliana)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257</th>\n",
       "      <td>Sobic.010G249200</td>\n",
       "      <td>GATA</td>\n",
       "      <td>MP00369</td>\n",
       "      <td>Sorghum bicolor</td>\n",
       "      <td>DAP</td>\n",
       "      <td>PlantTFDB</td>\n",
       "      <td>transfer from AT3G21175(Arabidopsis thaliana)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258</th>\n",
       "      <td>Sobic.010G254300</td>\n",
       "      <td>G2-like</td>\n",
       "      <td>MP00354</td>\n",
       "      <td>Sorghum bicolor</td>\n",
       "      <td>DAP</td>\n",
       "      <td>PlantTFDB</td>\n",
       "      <td>transfer from AT3G13040(Arabidopsis thaliana)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>Sobic.010G261800</td>\n",
       "      <td>MIKC_MADS</td>\n",
       "      <td>MP00077</td>\n",
       "      <td>Sorghum bicolor</td>\n",
       "      <td>ChIP-seq</td>\n",
       "      <td>PlantTFDB</td>\n",
       "      <td>transfer from AT3G54340(Arabidopsis thaliana)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>260 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Gene_id     Family Matrix_id          Species    Method  \\\n",
       "0    Sobic.001G002500        ERF   MP00374  Sorghum bicolor       DAP   \n",
       "1    Sobic.001G007200        HSF   MP00514  Sorghum bicolor       DAP   \n",
       "2    Sobic.001G026500        SBP   MP00633  Sorghum bicolor       PBM   \n",
       "3    Sobic.001G034300        Dof   MP00295  Sorghum bicolor       DAP   \n",
       "4    Sobic.001G036800        AP2   MP00615  Sorghum bicolor       PBM   \n",
       "..                ...        ...       ...              ...       ...   \n",
       "255  Sobic.010G215700        SBP   MP00090  Sorghum bicolor     SELEX   \n",
       "256  Sobic.010G236300        ARF   MP00461  Sorghum bicolor       DAP   \n",
       "257  Sobic.010G249200       GATA   MP00369  Sorghum bicolor       DAP   \n",
       "258  Sobic.010G254300    G2-like   MP00354  Sorghum bicolor       DAP   \n",
       "259  Sobic.010G261800  MIKC_MADS   MP00077  Sorghum bicolor  ChIP-seq   \n",
       "\n",
       "    Datasource                                  Datasource_ID  \n",
       "0    PlantTFDB  transfer from AT3G23240(Arabidopsis thaliana)  \n",
       "1    PlantTFDB  transfer from AT5G16820(Arabidopsis thaliana)  \n",
       "2    PlantTFDB       transfer from PK06791.1(Cannabis sativa)  \n",
       "3    PlantTFDB  transfer from AT2G37590(Arabidopsis thaliana)  \n",
       "4    PlantTFDB  transfer from AT2G28550(Arabidopsis thaliana)  \n",
       "..         ...                                            ...  \n",
       "255  PlantTFDB  transfer from AT1G02065(Arabidopsis thaliana)  \n",
       "256  PlantTFDB  transfer from AT4G30080(Arabidopsis thaliana)  \n",
       "257  PlantTFDB  transfer from AT3G21175(Arabidopsis thaliana)  \n",
       "258  PlantTFDB  transfer from AT3G13040(Arabidopsis thaliana)  \n",
       "259  PlantTFDB  transfer from AT3G54340(Arabidopsis thaliana)  \n",
       "\n",
       "[260 rows x 7 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in the plantregdb TFs in sorghum\n",
    "\n",
    "# plant TF db description\n",
    "df_tf_info = pd.read_csv('../data/plantregdb/tfs/Sbi_TF_binding_motifs_information.txt', delimiter='\\t')\n",
    "df_tf_info['origin'] = df_tf_info.apply(\n",
    "    lambda r: r.Datasource_ID.split('(')[1][:-1],\n",
    "    axis=1\n",
    ")\n",
    "\n",
    "# create df of tf motif\n",
    "meme_dir = '../data/plantregdb/tfs/Sbi_TF_binding_motifs_individual/'\n",
    "meme_files = [f for f in listdir(meme_dir) if f.endswith('.meme')]\n",
    "\n",
    "tf_to_motif = {}\n",
    "for f in meme_files:\n",
    "    tf = f[:-5]\n",
    "    motif =get_motif_from_meme_file(meme_dir + f)\n",
    "    tf_to_motif[tf] = motif\n",
    "\n",
    "print('unique motifs:', len(set(list(tf_to_motif.values())))) # there is only 242 unique motifs\n",
    "\n",
    "df_motif_seqs = pd.DataFrame.from_dict(tf_to_motif, columns= ['motif_seq'], orient='index')\n",
    "df_tf_db = df_tf_info.merge(df_motif_seqs, left_on='Gene_id', right_index=True)\n",
    "\n",
    "# reduce set of motifs to motifs that contain other smaller motifs.\n",
    "db_ms = set(list(tf_to_motif.values()))\n",
    "db_ms_unique = reduce_motifs(db_ms)\n",
    "small_list2 = reduce_motifs(db_ms_unique)\n",
    "# create a dictionary of original motif to uber motif\n",
    "om_to_um = {} # original motif to uber motif\n",
    "for m in db_ms:\n",
    "    ums = [k for k in db_ms_unique if m in k]\n",
    "    om_to_um[m] = ums\n",
    "\n",
    "df_tf_db['uber_motifs'] = df_tf_db.apply(lambda r: om_to_um[r.motif_seq], axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## find expression values of transcription factor binding sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "260"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# prioritize based on expression of TF in sorghum leaf from: http://sorghum.riken.jp/Downloads.html\n",
    "\n",
    "df_fpkm = pd.read_csv('../data/sorghum_riken/tf_fpkm.csv')\n",
    "leaf_stringent = [\n",
    "'Leaf', 'Keller-leaf','leaf_blade', 'leaf_flag_internode', 'leaf_lower',\n",
    "       'leaf_lower_growing', 'leaf_lower_whorl', 'leaf_middle_whorl',\n",
    "       'leaf_sheath_growing', 'leaf_upper', 'leaf_upper_growing',\n",
    "       'leaf_upper_whorl'\n",
    "]\n",
    "df_fpkm['mean_fpkm'] = df_fpkm[leaf_stringent].apply(lambda r: sum(r.values)/len(leaf_stringent), axis=1)\n",
    "\n",
    "df_tf_db= df_tf_db.merge(df_fpkm[['mean_fpkm', 'gene']], left_on = 'Gene_id', right_on = 'gene')\n",
    "\n",
    "# select genes that have non zero fpkm in either of the leaf observations. \n",
    "#better to take too many than too little\n",
    "# select genes that are not arabidopsis\n",
    "df_db_select = df_tf_db.loc[\n",
    "    (df_tf_db.mean_fpkm > 2) |\n",
    "    (df_tf_db.origin != 'Arabidopsis thaliana')\n",
    "]\n",
    "\n",
    "# get unique motifs\n",
    "uber_motifs_db = []\n",
    "for um_list in df_db_select.uber_motifs:\n",
    "    for um in um_list:\n",
    "        uber_motifs_db.append(um)\n",
    "uber_motfs_db_unique = set(uber_motifs_db)\n",
    "print(len(uber_motfs_db_unique))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TF motifs from DNase I hypersensitivity sites\n",
    "from: https://academic.oup.com/plcell/article/31/10/2297/5985706\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{\"3'UTR\", \"5'UTR\", 'Downstream', 'Exon/CD', 'Intron', 'Upstream'}"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_bur = pd.read_csv('../data/burgess2019/TPC2019-LSB-00078DR1-A_Supplemental_Data_set_9_p2.csv')\n",
    "df_sb = df_bur.loc[df_bur['sample'] == 'sbwl']\n",
    "\n",
    "# select motifs in upstream and 5'utr\n",
    "df_sb_up = df_sb.loc[df_sb.genomic_feature.isin(['Upstream', \"5'UTR\"])]\n",
    "\n",
    "# select significant motifs only\n",
    "df_sb_up_sig = df_sb_up.loc[df_sb_up.q_value<0.1]\n",
    "\n",
    "# convert all sequences to uppercase\n",
    "df_sb_up_sig['seq_upper'] = [k.upper() for k in df_sb_up_sig.sequence]\n",
    "\n",
    "# get a set of unique 'ubermotifs', ie. longer motifs that contain all the shorter ones.\n",
    "burgess_ums = reduce_motifs(df_sb_up_sig.seq_upper)\n",
    "\n",
    "# add a column of ubermotif\n",
    "\n",
    "om_to_um = {} # original motif to uber motif\n",
    "for m in df_sb_up_sig.seq_upper:\n",
    "    ums = [k for k in burgess_ums if m in k]\n",
    "    om_to_um[m] = ums\n",
    "\n",
    "df_sb_up_sig['uber_motifs'] = df_sb_up_sig.apply(\n",
    "    lambda r: om_to_um[r.seq_upper], \n",
    "    axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# get orthogonal primers for PCR amplification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get primers that enable orthogonal amplification of the library\n",
    "df_primers_subu = pd.read_csv('../data/ortho_primers/ysx008_supp_st_1.csv')\n",
    "df_primers_choose = df_primers_subu[df_primers_subu['Keep primer in orthogonal set?'] == 'Yes'] "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# generate all motif sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "217"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_motifs_to_insert = list(burgess_ums) + list(uber_motfs_db_unique)\n",
    "\n",
    "\n",
    "n_to_m_ins = dict(zip(range(len(all_motifs_to_insert)), all_motifs_to_insert))\n",
    "len(all_motifs_to_insert)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# generate mutations\n",
    "\n",
    "using BSAI for all the inserts, \n",
    "for the backbone primeres: \n",
    "- use BSAI for all the synthetic GFP driving constructs (coded starting with 2)\n",
    "- For full length native genes due to existing Type II restriction sites in the gene:\n",
    "    - 3A SBPase: use BsmBI\n",
    "    - 3B Raf1: use BsaI or AarI\n",
    "    - 3C PsbS: use AarI\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "current segment to mutate: GTCCGGAACCATTTGCTCTAAAGAACAGAATGGAACACCGGAATATCCAACGAATCCCTTGTAATTGTACCGCTATGAAATAGTAAATCCTCTTCTTCTTTTTTTTTGGCAAAGGGAATGAAAATAAATCTAACTTAGGCCTTGTTTAGTTGGCAAAAACTTTGGTTTCTGCCTACTGTAGCACTTTCGTTTTTATTTGA\n",
      "len of current segment to mtuate 200\n",
      "gene psbs, segment -1168, singles #: 600\n",
      "gene psbs, segment-1168, 2bp dels #: 199\n",
      "gene psbs, segment -1168, 12bp dels #: 189\n",
      "gene psbs, segment -1168, abe #: 192\n",
      "1\n",
      "current segment to mutate: CAAACATTGTCCAATCATGGAGTAACTAGGCTCAAAAGATTCATCTCACAAATTACAGTTAAACTGTGCAATTAGTTTTTATTTTTATCTATATTTAATGCTCCATGCATGCGATCAAAGATTTGATGTGACGGAGAATCTTGAAAATTTTTGCGAACTAAACAAGGCCTTACAGCAACATTGCCACAGCGGGTTTTTTA\n",
      "len of current segment to mtuate 200\n",
      "gene psbs, segment -968, singles #: 600\n",
      "gene psbs, segment-968, 2bp dels #: 199\n",
      "gene psbs, segment -968, 12bp dels #: 189\n",
      "gene psbs, segment -968, abe #: 188\n",
      "2\n",
      "current segment to mutate: CTTGGTGAAGAGTTTTCCTTCAAAATAGACCTTCAAGGTTCAGAACTTCAGATACAAAAATATACTTACTTTGCAGTGGAGGTTGAAATTTCTTAGCAAGATCCAATTCATCGACTTGAGAGAAACGAAAGAGCACGATGCTACGATCTGCTCCATACATGACAACCGTCACTAACAGACAGTAGAATGGACTCCTAGCA\n",
      "len of current segment to mtuate 200\n",
      "gene psbs, segment -768, singles #: 600\n",
      "gene psbs, segment-768, 2bp dels #: 199\n",
      "gene psbs, segment -768, 12bp dels #: 189\n",
      "gene psbs, segment -768, abe #: 176\n",
      "3\n",
      "current segment to mutate: ACTTCATTCACACATGCCTCAAGGCTGGAGGGGCAGAACAGAACACTTTTTGCTTGACGCGTGTATCTCACCGACCGACCAGCGCCATCAGTTTGTTGCGGCGCCTGACGCCCTCCCCCAATGCCTGCGTCTCCAGCCGTACGGAGGTCGCAGGACAGGAGCCGCACGCACGCTGGCTCCTGGGCTCCCTGGCCTCGGCG\n",
      "len of current segment to mtuate 200\n",
      "gene psbs, segment -568, singles #: 600\n",
      "gene psbs, segment-568, 2bp dels #: 199\n",
      "gene psbs, segment -568, 12bp dels #: 189\n",
      "gene psbs, segment -568, abe #: 79\n",
      "4\n",
      "current segment to mutate: CGGCGCGGCCATAGATTGGATTCTCGGCGACGTGGCCTCCTGGCAACCCATCCCGCAACGGCGCACCCTATCCTCCGCCTCGCCGCTCGCCAGTCGCCAATCGACACCTCCAGTTGGCGCCACTCCAGCTTCACGCTGCCCGCGCCTGTGACCCCAATGCCCCCTTTCATAAGAACGGGACACGCGGCCTCGGAGCCAGT\n",
      "len of current segment to mtuate 200\n",
      "gene psbs, segment -368, singles #: 600\n",
      "gene psbs, segment-368, 2bp dels #: 199\n",
      "gene psbs, segment -368, 12bp dels #: 189\n",
      "gene psbs, segment -368, abe #: 73\n",
      "5\n",
      "doing last segment\n",
      "current segment to mutate: AGACACGGGAGCCCACGACCACGCAGCTCGCAGCCGCGCGCGCTCGCCTATACGACTACGAGTACACAGCTTCGATCCTCAGCGCGCGCGAGTACGTGGCTGCTGCGAATTCCTATCTTTCTCTGTGTTCGCGTGTTCGGTGTTCGTGAGTAGGCGCGGTCAGCACAT\n",
      "len of current segment to mtuate 168\n",
      "gene psbs, segment -168, singles #: 504\n",
      "gene psbs, segment-168, 2bp dels #: 167\n",
      "gene psbs, segment -168, 12bp dels #: 157\n",
      "gene psbs, segment -168, abe #: 83\n",
      "6\n",
      "number of singles, abes, 2+10bp deletions: 6559\n",
      "motif_chunk length for motif inserts 199\n",
      "7\n",
      "current segment to mutate: AAAAGTTAGACAAAGAAGATAATAATTATAAGTTAATATCTACTAGTTACACACTTACACGTAGAAAGTTAGATATAAAAGTTGCGTACAGAAAGTTAGACATAAAAGTTACGTGCAGAAAGTTAATAATTACAAATTATATGGTTAAAATTTACATCTAGCATTTTAAAATAAATGTTTTACTTTTTTATGACATCAGC\n",
      "len of current segment to mtuate 200\n",
      "gene raf1, segment -1050, singles #: 600\n",
      "gene raf1, segment-1050, 2bp dels #: 199\n",
      "gene raf1, segment -1050, 12bp dels #: 189\n",
      "gene raf1, segment -1050, abe #: 246\n",
      "8\n",
      "current segment to mutate: GGTGGTTATGGAAAGGCGAAGGAGAGGTAGTTTGAACCTTTGTTACATCAAATCGTGAACGATCACCCAAATAAAATAAGGGTTGGCAGCATATATAAAAAGAAAAAATTGATAGACTTGTATAAAAAATAAAAAGAGAAAAAAGACAAAAGAAAATTTAAAAAAATGTTTTATAATTCTGTGTTTTAAAGTTTTGGCAT\n",
      "len of current segment to mtuate 200\n",
      "gene raf1, segment -850, singles #: 600\n",
      "gene raf1, segment-850, 2bp dels #: 199\n",
      "gene raf1, segment -850, 12bp dels #: 189\n",
      "gene raf1, segment -850, abe #: 205\n",
      "9\n",
      "current segment to mutate: AAGTTGTAAGTTATATGGTATAAGTCATATGGTGATAAGTTTTGCAAAGTGTTATACCAAGTTCTTAGTATAATTTATAGAAATAAGTTATGTTCTATAACATATATGTATAACTTAGTTACCTTCTAAAAATGAAAAGTTGCGGAAATTATTATTTCCAAATAAGAAAAGTTGCGGGAGTAGCTGATTCCTTACTGATG\n",
      "len of current segment to mtuate 200\n",
      "gene raf1, segment -650, singles #: 600\n",
      "gene raf1, segment-650, 2bp dels #: 199\n",
      "gene raf1, segment -650, 12bp dels #: 189\n",
      "gene raf1, segment -650, abe #: 232\n",
      "10\n",
      "current segment to mutate: AACGATCGTAGAATTGGATTTGTGTCTATGATTTTGCCTTGCGACTATTAAACATGATGCTTGGACTGTGTTATGCTTTTTGAAATAAAAAAGTAGGAGAAACCTCCCTTTCTTAAATAAAAAAACTTCAAAATCATATATTCATAAAAACTTCCTACAGGCAACAAATAATATCTATCAAGTTGCGTATCTGAATTTTT\n",
      "len of current segment to mtuate 200\n",
      "gene raf1, segment -450, singles #: 600\n",
      "gene raf1, segment-450, 2bp dels #: 199\n",
      "gene raf1, segment -450, 12bp dels #: 189\n",
      "gene raf1, segment -450, abe #: 213\n",
      "11\n",
      "current segment to mutate: ATATAGGAAAATTCAAACCTCGAAGAAGAGAAGCGTCCGTTGAGTGGAGAGAGAAGGCTCGATTCAGAAGCAAGCGCGACAGAAACGGAAAGTACCTGACGGAAAGGGAAACGGAAGGAAAAAACAGGAGTGCGCGTCGGTAGCGCCCGTGTGGCGCACGCTAGGACAAGGATAAGCCCTGGTTCCGTTCCGTGCTTCCC\n",
      "len of current segment to mtuate 200\n",
      "gene raf1, segment -250, singles #: 600\n",
      "gene raf1, segment-250, 2bp dels #: 199\n",
      "gene raf1, segment -250, 12bp dels #: 189\n",
      "gene raf1, segment -250, abe #: 134\n",
      "12\n",
      "doing last segment\n",
      "current segment to mutate: AGACTCTCATTCCTCTTCTCCCCTCCCCGATCCCCCCCCTTATCGCCACC\n",
      "len of current segment to mtuate 50\n",
      "gene raf1, segment -50, singles #: 150\n",
      "gene raf1, segment-50, 2bp dels #: 49\n",
      "gene raf1, segment -50, 12bp dels #: 39\n",
      "gene raf1, segment -50, abe #: 17\n",
      "13\n",
      "number of singles, abes, 2+10bp deletions: 6225\n",
      "motif_chunk length for motif inserts 199\n",
      "14\n",
      "current segment to mutate: TCATCAGTCAAAGCGGCACTTCGCTGAGCTGAGATAGAGAAGATGACACGGTCTTGCTACCACAAAAAAGAGGACACCCTTCTCCTTATGCACATGAAACCATCACGGTGAGGTCCTCTAGTCTTGCTAGCGCCACTAGGTTGATGAGGTGCAGAAACGATCTTCTATGACACAATGGCATCTAGAACTGGCTCGCTAGT\n",
      "len of current segment to mtuate 200\n",
      "gene zmm28, segment -1127, singles #: 600\n",
      "gene zmm28, segment-1127, 2bp dels #: 199\n",
      "gene zmm28, segment -1127, 12bp dels #: 189\n",
      "gene zmm28, segment -1127, abe #: 135\n",
      "15\n",
      "current segment to mutate: GGCAGTGCTCGTGAAGTGGAGGAACAAATAGAAGCTTGAGGAAGTGTCGATGCACGATGAGGAGCAAAAATTGTCCATGATAATTGTGCATGTTTTTATTTGCTCTCTTGGAGAAATAAACTACTAGGTTTACATTTTAGCTATTTTTCTAGCTCTTGCTACAATGTTACTTTCACCCTTCGTGAAAAAAATCACTTGAG\n",
      "len of current segment to mtuate 200\n",
      "gene zmm28, segment -927, singles #: 600\n",
      "gene zmm28, segment-927, 2bp dels #: 199\n",
      "gene zmm28, segment -927, 12bp dels #: 189\n",
      "gene zmm28, segment -927, abe #: 175\n",
      "16\n",
      "current segment to mutate: ATATTTTGATTAGAATCTGTATTTGTATTTGTATTTGTATGCTCTCTTGGAGTAATAGTGTAGTATGCTTGATATGTAAGCAAACCAAAAAAAGAGTCAATTTCGGCAACCAATTTGGGGATGTTGCAGTCCAGAAATGGAAGAAAAAGGGCAGAGCACTCACGGACATGCGGTCCCACATCCGTGTGGACCCATCAGCC\n",
      "len of current segment to mtuate 200\n",
      "gene zmm28, segment -727, singles #: 600\n",
      "gene zmm28, segment-727, 2bp dels #: 199\n",
      "gene zmm28, segment -727, 12bp dels #: 189\n",
      "gene zmm28, segment -727, abe #: 166\n",
      "17\n",
      "current segment to mutate: AGTCCCCCTGAAGCTACCCCACGTGTCACTTGCCACCTCGCTCCCTTGTGCCCGACAGGCACAGGGTCAGGGAAAACCATTTCTCACCGCGGATCCGCGCTACCCACAGCCGCACCCTTTCCGGCCCACGACGGGCTCGACCAATCCCTGCCCCGCGCGCCCTGACGAGCGCAGCACTCCCACGCCGGCCTCGCTGGGCC\n",
      "len of current segment to mtuate 200\n",
      "gene zmm28, segment -527, singles #: 600\n",
      "gene zmm28, segment-527, 2bp dels #: 199\n",
      "gene zmm28, segment -527, 12bp dels #: 189\n",
      "gene zmm28, segment -527, abe #: 63\n",
      "18\n",
      "current segment to mutate: CATCCTTCTGGCCCATAATAACGATCCCCCTCGGGATCCGACGGTCCATCTGCCTCCACGCCGCTCCGAAACCTCATCGTCCAATCAAAACACGACAGCGGGTCAGGCAAAACCACCGTGGTTTCGCGATACCGCTCTCCTCCCCATCTAAAACCGCCCACCTCCTCTTCCTCTCCTCCTCCTAGCTCTCTCTTGTCTGC\n",
      "len of current segment to mtuate 200\n",
      "gene zmm28, segment -327, singles #: 600\n",
      "gene zmm28, segment-327, 2bp dels #: 199\n",
      "gene zmm28, segment -327, 12bp dels #: 189\n",
      "gene zmm28, segment -327, abe #: 105\n",
      "19\n",
      "doing last segment\n",
      "current segment to mutate: GCACCACTCGAGTCGAGAGACGAAAACCACCTGATCCTTCCTTCCCCGTCTTCCTCCTCCCACACGTCCCCATCGCCATTGCTGCGCCGAGAGTGAGCGGGAGAGGGTAGGTGTCGAGGCGGCGGAG\n",
      "len of current segment to mtuate 127\n",
      "gene zmm28, segment -127, singles #: 381\n",
      "gene zmm28, segment-127, 2bp dels #: 126\n",
      "gene zmm28, segment -127, 12bp dels #: 116\n",
      "gene zmm28, segment -127, abe #: 51\n",
      "20\n",
      "number of singles, abes, 2+10bp deletions: 6258\n",
      "motif_chunk length for motif inserts 199\n",
      "21\n",
      "current segment to mutate: TTCTGCGAGGCCCAAGTCAAACATTTTTTTTCACAAATACATAAAATATTTGTATATCATTGCAATTAGGTAGAAGAGTTTAACCATACAGAAAAAAACACGCTTCTGGCGAGCATCATTGTAATGAAGTCAAACATTTCGTTTGGACTTTAATAGCCAGTGTTGCTCTTTTAATGTCTCCATTTTTTGTAAGTGTCATA\n",
      "len of current segment to mtuate 200\n",
      "gene sbpase, segment -1257, singles #: 600\n",
      "gene sbpase, segment-1257, 2bp dels #: 199\n",
      "gene sbpase, segment -1257, 12bp dels #: 189\n",
      "gene sbpase, segment -1257, abe #: 201\n",
      "22\n",
      "current segment to mutate: TTTGGTTTGTCCTAAGTCAAACATTGAAATACTTGACCAACCATTTATAATTAATAATCTATTTAGATAATATATGCATTAGATATTATGATAGTATTTCTCATAATAAATTTAAAAATATAAATTTTATAACATGAAATTATATAAATATTTTGAAATCAATAGTCAAAGAGACACACATTTGACTTAGAAACAAATTA\n",
      "len of current segment to mtuate 200\n",
      "gene sbpase, segment -1057, singles #: 600\n",
      "gene sbpase, segment-1057, 2bp dels #: 199\n",
      "gene sbpase, segment -1057, 12bp dels #: 189\n",
      "gene sbpase, segment -1057, abe #: 269\n",
      "23\n",
      "current segment to mutate: AATGTGACACTTGAAAAAAAACAAAGGGAGTAATTCTGTCTGGGCAGGCTTGAACTCCTGCTCATGTTATGACTCAAAGTTGCTAAAGATGTCAAGTGGCAATGTATGTGCAGACAAGCTTATGTTCTTAAGCTTTTTATATAAGCATAAGCAGCAAATGCTTTGCAAAGAACCGACTTAAAACATCACACTGAACAATT\n",
      "len of current segment to mtuate 200\n",
      "gene sbpase, segment -857, singles #: 600\n",
      "gene sbpase, segment-857, 2bp dels #: 199\n",
      "gene sbpase, segment -857, 12bp dels #: 189\n",
      "gene sbpase, segment -857, abe #: 192\n",
      "24\n",
      "current segment to mutate: TAGAAAAGGATCCATACCGATACGCCGAAACCGTCGTACTCGTATATATACGACAATTAAATACTAGAACCAAACATCCCAGACCGAATAATGGGAAACGAGCCAGTTTCTGCGACTTAAACATCACACTGAAGTACTGAACAATTCAGAAAAAGTTCCATACCGATACGGCGATACCGTCGTACTCGTATATACGACAA\n",
      "len of current segment to mtuate 200\n",
      "gene sbpase, segment -657, singles #: 600\n",
      "gene sbpase, segment-657, 2bp dels #: 199\n",
      "gene sbpase, segment -657, 12bp dels #: 189\n",
      "gene sbpase, segment -657, abe #: 178\n",
      "25\n",
      "current segment to mutate: TTAAATACCAGAACCGAACAATTTCAGAAAAAATGATCCATATCGTATATACGACAATTAAATACCAGAACAGAGCCAAACAAACATCCTAGACCGAAATGGGCCTGTTTCTGGTTGCTTGGAGTTGGAGTTGGAGTGCGTACTGCGTAGGCCCGTCCGATCCGTGGCTCATCTCGGCATGGCCACACAAACCACGAATG\n",
      "len of current segment to mtuate 200\n",
      "gene sbpase, segment -457, singles #: 600\n",
      "gene sbpase, segment-457, 2bp dels #: 199\n",
      "gene sbpase, segment -457, 12bp dels #: 189\n",
      "gene sbpase, segment -457, abe #: 151\n",
      "26\n",
      "current segment to mutate: GTCATCGTGCACCGCAGCGCGGCCCTCCCCTTTCTATCTTCTCCAGCTCCAAATGGCGCCACCGCGGCGGCCGGGTTATCTTGTCCAGACGTGCATCCCTCCGTTGTGATCCATTCCTCTCCGCGCCGCGATAGCTAGCTTAGTTGCAAGCACCAGCAGCTAGCTGACCACCAAAACGTTAGCGCACGCAGACAATTTCG\n",
      "len of current segment to mtuate 200\n",
      "gene sbpase, segment -257, singles #: 600\n",
      "gene sbpase, segment-257, 2bp dels #: 199\n",
      "gene sbpase, segment -257, 12bp dels #: 189\n",
      "gene sbpase, segment -257, abe #: 102\n",
      "27\n",
      "doing last segment\n",
      "current segment to mutate: ACAGCATCCATCAGGTTTTAGGCCCCTGCTGCCGTCGTTGAGCGCCACCTGAAAGCA\n",
      "len of current segment to mtuate 57\n",
      "gene sbpase, segment -57, singles #: 171\n",
      "gene sbpase, segment-57, 2bp dels #: 56\n",
      "gene sbpase, segment -57, 12bp dels #: 46\n",
      "gene sbpase, segment -57, abe #: 23\n",
      "28\n",
      "number of singles, abes, 2+10bp deletions: 7317\n",
      "motif_chunk length for motif inserts 199\n",
      "29\n"
     ]
    }
   ],
   "source": [
    "c=0\n",
    "df_muts_psbsb, c = get_gene_library(\n",
    "    dic_dic_gene_utr['psbs']['prime_5'], \n",
    "    dic_dic_gene_utr['psbs']['utr_5'], \n",
    "    dic_dic_gene_utr['psbs']['exon_1'], \n",
    "    c, \n",
    "    'psbs', \n",
    "    n_to_m_ins\n",
    ")\n",
    "\n",
    "df_muts_raf1, c = get_gene_library(\n",
    "    dic_dic_gene_utr['raf1']['prime_5'], \n",
    "    dic_dic_gene_utr['raf1']['utr_5'], \n",
    "    dic_dic_gene_utr['raf1']['exon_1'], \n",
    "    c, 'raf1', \n",
    "    n_to_m_ins)\n",
    "\n",
    "# these were ordered but fell out during cloning\n",
    "df_muts_zmm28, c = get_gene_library(\n",
    "    dic_dic_gene_utr['zmm28']['prime_5'], \n",
    "    dic_dic_gene_utr['zmm28']['utr_5'], \n",
    "    dic_dic_gene_utr['zmm28']['exon_1'], \n",
    "    c, 'zmm28', n_to_m_ins)\n",
    "\n",
    "df_muts_sbpase, c = get_gene_library(\n",
    "    dic_dic_gene_utr['sbpase']['prime_5'], \n",
    "    dic_dic_gene_utr['sbpase']['utr_5'], \n",
    "    dic_dic_gene_utr['sbpase']['exon_1'], \n",
    "    c, 'sbpase', n_to_m_ins)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# filtering down to get fewer insertions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "26359"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n_muts_non_ins = (len(df_muts_sbpase.loc[~df_muts_sbpase['Unnamed: 0'].str.startswith('ins')]) + \n",
    "                    len(df_muts_zmm28.loc[~df_muts_zmm28['Unnamed: 0'].str.startswith('ins')]) + \n",
    "                    len(df_muts_raf1.loc[~df_muts_raf1['Unnamed: 0'].str.startswith('ins')]) + \n",
    "                  len(df_muts_psbsb.loc[~df_muts_psbsb['Unnamed: 0'].str.startswith('ins')])\n",
    "                 )\n",
    "n_muts_non_ins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_motifs_to_insert = list(burgess_ums) + list(uber_motfs_db_unique)\n",
    "n_to_m_ins = dict(zip(range(len(all_motifs_to_insert)), all_motifs_to_insert))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3034\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "35"
      ]
     },
     "execution_count": 138,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get the top35 most frequently occuring motifs from Burgess et al. and the rest of the motifs\n",
    "uber_motifs_occurences = []\n",
    "for n,r in df_sb_up_sig.iterrows():\n",
    "    uber_motifs_occurences += r.uber_motifs\n",
    "\n",
    "um_burgess_counts = sorted(Counter(uber_motifs_occurences).items(), key=lambda x: x[1], reverse=True)\n",
    "\n",
    "um_burgess_all = um_burgess_counts[:35]\n",
    "um_burgess_split = um_burgess_counts[35:]\n",
    "\n",
    "m_all = [k[0] for k in um_burgess_all]\n",
    "m_split = [k[0] for k in um_burgess_split] + list(uber_motfs_db_unique)\n",
    "\n",
    "n_motif_all = [n for n,m in n_to_m_ins.items() if m in m_all]\n",
    "n_motif_split = [n for n,m in n_to_m_ins.items() if m in m_split]\n",
    "\n",
    "\n",
    "chunk_size = int(len(n_motif_split)/4)\n",
    "chunked_list = [n_motif_split[i:i+chunk_size] for i in range(0, len(n_motif_split), chunk_size)]\n",
    "\n",
    "\n",
    "df_filtered_psbs = filter_df_muts(df_muts_psbsb, n_motif_all + chunked_list[0])\n",
    "df_filtered_raf1 = filter_df_muts(df_muts_raf1, n_motif_all + chunked_list[1])\n",
    "df_filtered_sbpase = filter_df_muts(df_muts_sbpase, n_motif_all + chunked_list[3])\n",
    "\n",
    "df_filtered_psbs['order_name'] = df_filtered_psbs['gene_name']  + df_filtered_psbs['Unnamed: 0'] \n",
    "df_filtered_raf1['order_name'] = df_filtered_raf1['gene_name']  + df_filtered_raf1['Unnamed: 0'] \n",
    "df_filtered_sbpase['order_name'] = df_filtered_sbpase['gene_name']  + df_filtered_sbpase['Unnamed: 0'] \n",
    "\n",
    "write_fasta_df_muts(df_filtered_psbs, 'df_filtered_psbs.fasta', fasta_name = 'order_name')\n",
    "write_fasta_df_muts(df_filtered_raf1, 'df_filtered_raf1.fasta', fasta_name = 'order_name')\n",
    "write_fasta_df_muts(df_filtered_sbpase, 'df_filtered_sbpase.fasta', fasta_name = 'order_name')\n",
    "\n",
    "df_all_order = pd.concat([df_filtered_psbs[['order_name', 'seq_order']], \n",
    "          df_filtered_raf1[['order_name', 'seq_order']], \n",
    "          df_filtered_zmm28[['order_name', 'seq_order']],\n",
    "          df_filtered_sbpase[['order_name', 'seq_order']]\n",
    "                            ])\n",
    "write_fasta_df_muts(df_all_order, 'df_all_order.fasta', fasta_name = 'order_name')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# order primers to clone out backbone and sublibraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_filtered_psbs = pd.read_csv('./df_filtered_psbs.csv')\n",
    "df_filtered_raf1 = pd.read_csv('./df_filtered_raf1.csv')\n",
    "df_filtered_zmm28 = pd.read_csv('./df_filtered_zmm28.csv')\n",
    "df_filtered_sbpase = pd.read_csv('./df_filtered_sbpase.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Unnamed: 0.1</th>\n",
       "      <th>min_chunk</th>\n",
       "      <th>bsa1_pre</th>\n",
       "      <th>bsa1_post</th>\n",
       "      <th>prime_5_add</th>\n",
       "      <th>amp_fwd</th>\n",
       "      <th>amp_rev</th>\n",
       "      <th>primer_ids</th>\n",
       "      <th>segment_loc</th>\n",
       "      <th>seq_order</th>\n",
       "      <th>len_seq_order</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>single_G-1168A</td>\n",
       "      <td>ATCCGGAACCATTTGCTCTAAAGAACAGAATGGAACACCGGAATAT...</td>\n",
       "      <td>GGTCTCCCATTG</td>\n",
       "      <td>CAAACAGAGACC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AAACACGTGGCAAACATTCC</td>\n",
       "      <td>GTACTGTATGGCTCCGGTTT</td>\n",
       "      <td>1_2</td>\n",
       "      <td>-1168</td>\n",
       "      <td>AAACACGTGGCAAACATTCCGGTCTCCCATTGATCCGGAACCATTT...</td>\n",
       "      <td>264</td>\n",
       "      <td>psbs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>single_G-1168C</td>\n",
       "      <td>CTCCGGAACCATTTGCTCTAAAGAACAGAATGGAACACCGGAATAT...</td>\n",
       "      <td>GGTCTCCCATTG</td>\n",
       "      <td>CAAACAGAGACC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AAACACGTGGCAAACATTCC</td>\n",
       "      <td>GTACTGTATGGCTCCGGTTT</td>\n",
       "      <td>1_2</td>\n",
       "      <td>-1168</td>\n",
       "      <td>AAACACGTGGCAAACATTCCGGTCTCCCATTGCTCCGGAACCATTT...</td>\n",
       "      <td>264</td>\n",
       "      <td>psbs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>single_G-1168T</td>\n",
       "      <td>TTCCGGAACCATTTGCTCTAAAGAACAGAATGGAACACCGGAATAT...</td>\n",
       "      <td>GGTCTCCCATTG</td>\n",
       "      <td>CAAACAGAGACC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AAACACGTGGCAAACATTCC</td>\n",
       "      <td>GTACTGTATGGCTCCGGTTT</td>\n",
       "      <td>1_2</td>\n",
       "      <td>-1168</td>\n",
       "      <td>AAACACGTGGCAAACATTCCGGTCTCCCATTGTTCCGGAACCATTT...</td>\n",
       "      <td>264</td>\n",
       "      <td>psbs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>single_T-1167A</td>\n",
       "      <td>GACCGGAACCATTTGCTCTAAAGAACAGAATGGAACACCGGAATAT...</td>\n",
       "      <td>GGTCTCCCATTG</td>\n",
       "      <td>CAAACAGAGACC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AAACACGTGGCAAACATTCC</td>\n",
       "      <td>GTACTGTATGGCTCCGGTTT</td>\n",
       "      <td>1_2</td>\n",
       "      <td>-1168</td>\n",
       "      <td>AAACACGTGGCAAACATTCCGGTCTCCCATTGGACCGGAACCATTT...</td>\n",
       "      <td>264</td>\n",
       "      <td>psbs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>single_T-1167C</td>\n",
       "      <td>GCCCGGAACCATTTGCTCTAAAGAACAGAATGGAACACCGGAATAT...</td>\n",
       "      <td>GGTCTCCCATTG</td>\n",
       "      <td>CAAACAGAGACC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AAACACGTGGCAAACATTCC</td>\n",
       "      <td>GTACTGTATGGCTCCGGTTT</td>\n",
       "      <td>1_2</td>\n",
       "      <td>-1168</td>\n",
       "      <td>AAACACGTGGCAAACATTCCGGTCTCCCATTGGCCCGGAACCATTT...</td>\n",
       "      <td>264</td>\n",
       "      <td>psbs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13712</th>\n",
       "      <td>24668</td>\n",
       "      <td>ins_motif_m212_p195_rev</td>\n",
       "      <td>ACGACAATTAAATACCAGAACAGAGCCAAACAAACATCCTAGACCG...</td>\n",
       "      <td>GGTCTCGTATAT</td>\n",
       "      <td>CAAATGGAGACC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGGACACCAGACCAATGAAG</td>\n",
       "      <td>CGCTGATGGTAATTAGCCCT</td>\n",
       "      <td>62_63</td>\n",
       "      <td>-150</td>\n",
       "      <td>AGGACACCAGACCAATGAAGGGTCTCGTATATACGACAATTAAATA...</td>\n",
       "      <td>278</td>\n",
       "      <td>sbpase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13713</th>\n",
       "      <td>24669</td>\n",
       "      <td>ins_motif_m213_p195_fwd</td>\n",
       "      <td>ACGACAATTAAATACCAGAACAGAGCCAAACAAACATCCTAGACCG...</td>\n",
       "      <td>GGTCTCGTATAT</td>\n",
       "      <td>CAAATGGAGACC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGGACACCAGACCAATGAAG</td>\n",
       "      <td>CGCTGATGGTAATTAGCCCT</td>\n",
       "      <td>62_63</td>\n",
       "      <td>-150</td>\n",
       "      <td>AGGACACCAGACCAATGAAGGGTCTCGTATATACGACAATTAAATA...</td>\n",
       "      <td>272</td>\n",
       "      <td>sbpase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13714</th>\n",
       "      <td>24670</td>\n",
       "      <td>ins_motif_m213_p195_rev</td>\n",
       "      <td>ACGACAATTAAATACCAGAACAGAGCCAAACAAACATCCTAGACCG...</td>\n",
       "      <td>GGTCTCGTATAT</td>\n",
       "      <td>CAAATGGAGACC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGGACACCAGACCAATGAAG</td>\n",
       "      <td>CGCTGATGGTAATTAGCCCT</td>\n",
       "      <td>62_63</td>\n",
       "      <td>-150</td>\n",
       "      <td>AGGACACCAGACCAATGAAGGGTCTCGTATATACGACAATTAAATA...</td>\n",
       "      <td>272</td>\n",
       "      <td>sbpase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13715</th>\n",
       "      <td>24671</td>\n",
       "      <td>ins_motif_m214_p195_fwd</td>\n",
       "      <td>ACGACAATTAAATACCAGAACAGAGCCAAACAAACATCCTAGACCG...</td>\n",
       "      <td>GGTCTCGTATAT</td>\n",
       "      <td>CAAATGGAGACC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGGACACCAGACCAATGAAG</td>\n",
       "      <td>CGCTGATGGTAATTAGCCCT</td>\n",
       "      <td>62_63</td>\n",
       "      <td>-150</td>\n",
       "      <td>AGGACACCAGACCAATGAAGGGTCTCGTATATACGACAATTAAATA...</td>\n",
       "      <td>273</td>\n",
       "      <td>sbpase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13716</th>\n",
       "      <td>24672</td>\n",
       "      <td>ins_motif_m214_p195_rev</td>\n",
       "      <td>ACGACAATTAAATACCAGAACAGAGCCAAACAAACATCCTAGACCG...</td>\n",
       "      <td>GGTCTCGTATAT</td>\n",
       "      <td>CAAATGGAGACC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGGACACCAGACCAATGAAG</td>\n",
       "      <td>CGCTGATGGTAATTAGCCCT</td>\n",
       "      <td>62_63</td>\n",
       "      <td>-150</td>\n",
       "      <td>AGGACACCAGACCAATGAAGGGTCTCGTATATACGACAATTAAATA...</td>\n",
       "      <td>273</td>\n",
       "      <td>sbpase</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>51959 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Unnamed: 0             Unnamed: 0.1  \\\n",
       "0               0           single_G-1168A   \n",
       "1               1           single_G-1168C   \n",
       "2               2           single_G-1168T   \n",
       "3               3           single_T-1167A   \n",
       "4               4           single_T-1167C   \n",
       "...           ...                      ...   \n",
       "13712       24668  ins_motif_m212_p195_rev   \n",
       "13713       24669  ins_motif_m213_p195_fwd   \n",
       "13714       24670  ins_motif_m213_p195_rev   \n",
       "13715       24671  ins_motif_m214_p195_fwd   \n",
       "13716       24672  ins_motif_m214_p195_rev   \n",
       "\n",
       "                                               min_chunk      bsa1_pre  \\\n",
       "0      ATCCGGAACCATTTGCTCTAAAGAACAGAATGGAACACCGGAATAT...  GGTCTCCCATTG   \n",
       "1      CTCCGGAACCATTTGCTCTAAAGAACAGAATGGAACACCGGAATAT...  GGTCTCCCATTG   \n",
       "2      TTCCGGAACCATTTGCTCTAAAGAACAGAATGGAACACCGGAATAT...  GGTCTCCCATTG   \n",
       "3      GACCGGAACCATTTGCTCTAAAGAACAGAATGGAACACCGGAATAT...  GGTCTCCCATTG   \n",
       "4      GCCCGGAACCATTTGCTCTAAAGAACAGAATGGAACACCGGAATAT...  GGTCTCCCATTG   \n",
       "...                                                  ...           ...   \n",
       "13712  ACGACAATTAAATACCAGAACAGAGCCAAACAAACATCCTAGACCG...  GGTCTCGTATAT   \n",
       "13713  ACGACAATTAAATACCAGAACAGAGCCAAACAAACATCCTAGACCG...  GGTCTCGTATAT   \n",
       "13714  ACGACAATTAAATACCAGAACAGAGCCAAACAAACATCCTAGACCG...  GGTCTCGTATAT   \n",
       "13715  ACGACAATTAAATACCAGAACAGAGCCAAACAAACATCCTAGACCG...  GGTCTCGTATAT   \n",
       "13716  ACGACAATTAAATACCAGAACAGAGCCAAACAAACATCCTAGACCG...  GGTCTCGTATAT   \n",
       "\n",
       "          bsa1_post prime_5_add               amp_fwd               amp_rev  \\\n",
       "0      CAAACAGAGACC         NaN  AAACACGTGGCAAACATTCC  GTACTGTATGGCTCCGGTTT   \n",
       "1      CAAACAGAGACC         NaN  AAACACGTGGCAAACATTCC  GTACTGTATGGCTCCGGTTT   \n",
       "2      CAAACAGAGACC         NaN  AAACACGTGGCAAACATTCC  GTACTGTATGGCTCCGGTTT   \n",
       "3      CAAACAGAGACC         NaN  AAACACGTGGCAAACATTCC  GTACTGTATGGCTCCGGTTT   \n",
       "4      CAAACAGAGACC         NaN  AAACACGTGGCAAACATTCC  GTACTGTATGGCTCCGGTTT   \n",
       "...             ...         ...                   ...                   ...   \n",
       "13712  CAAATGGAGACC         NaN  AGGACACCAGACCAATGAAG  CGCTGATGGTAATTAGCCCT   \n",
       "13713  CAAATGGAGACC         NaN  AGGACACCAGACCAATGAAG  CGCTGATGGTAATTAGCCCT   \n",
       "13714  CAAATGGAGACC         NaN  AGGACACCAGACCAATGAAG  CGCTGATGGTAATTAGCCCT   \n",
       "13715  CAAATGGAGACC         NaN  AGGACACCAGACCAATGAAG  CGCTGATGGTAATTAGCCCT   \n",
       "13716  CAAATGGAGACC         NaN  AGGACACCAGACCAATGAAG  CGCTGATGGTAATTAGCCCT   \n",
       "\n",
       "      primer_ids  segment_loc  \\\n",
       "0            1_2        -1168   \n",
       "1            1_2        -1168   \n",
       "2            1_2        -1168   \n",
       "3            1_2        -1168   \n",
       "4            1_2        -1168   \n",
       "...          ...          ...   \n",
       "13712      62_63         -150   \n",
       "13713      62_63         -150   \n",
       "13714      62_63         -150   \n",
       "13715      62_63         -150   \n",
       "13716      62_63         -150   \n",
       "\n",
       "                                               seq_order  len_seq_order  \\\n",
       "0      AAACACGTGGCAAACATTCCGGTCTCCCATTGATCCGGAACCATTT...            264   \n",
       "1      AAACACGTGGCAAACATTCCGGTCTCCCATTGCTCCGGAACCATTT...            264   \n",
       "2      AAACACGTGGCAAACATTCCGGTCTCCCATTGTTCCGGAACCATTT...            264   \n",
       "3      AAACACGTGGCAAACATTCCGGTCTCCCATTGGACCGGAACCATTT...            264   \n",
       "4      AAACACGTGGCAAACATTCCGGTCTCCCATTGGCCCGGAACCATTT...            264   \n",
       "...                                                  ...            ...   \n",
       "13712  AGGACACCAGACCAATGAAGGGTCTCGTATATACGACAATTAAATA...            278   \n",
       "13713  AGGACACCAGACCAATGAAGGGTCTCGTATATACGACAATTAAATA...            272   \n",
       "13714  AGGACACCAGACCAATGAAGGGTCTCGTATATACGACAATTAAATA...            272   \n",
       "13715  AGGACACCAGACCAATGAAGGGTCTCGTATATACGACAATTAAATA...            273   \n",
       "13716  AGGACACCAGACCAATGAAGGGTCTCGTATATACGACAATTAAATA...            273   \n",
       "\n",
       "      gene_name  \n",
       "0          psbs  \n",
       "1          psbs  \n",
       "2          psbs  \n",
       "3          psbs  \n",
       "4          psbs  \n",
       "...         ...  \n",
       "13712    sbpase  \n",
       "13713    sbpase  \n",
       "13714    sbpase  \n",
       "13715    sbpase  \n",
       "13716    sbpase  \n",
       "\n",
       "[51959 rows x 13 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_all_order = pd.concat([df_filtered_psbs, \n",
    "          df_filtered_raf1, \n",
    "          df_filtered_zmm28,\n",
    "          df_filtered_sbpase\n",
    "                            ])\n",
    "df_all_order"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dic_pc_num_to_seq = dict(zip(df_primers_choose['Primer id'], df_primers_choose['Sequence']))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>p1</th>\n",
       "      <th>p2</th>\n",
       "      <th>p1_seq</th>\n",
       "      <th>p2_seq</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>AAACACGTGGCAAACATTCC</td>\n",
       "      <td>AAACCGGAGCCATACAGTAC</td>\n",
       "      <td>psbs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>AAAGCACTCTTAGGCCTCTG</td>\n",
       "      <td>AAAGGGGCCGTCAATATCAG</td>\n",
       "      <td>psbs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>AAATAAGACGACGACCCTCG</td>\n",
       "      <td>AACGATGATGCTCACTCTCG</td>\n",
       "      <td>psbs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11</td>\n",
       "      <td>12</td>\n",
       "      <td>AAGAATTACTGACCCCTCGG</td>\n",
       "      <td>AAGACGATCCGAGCCATTAC</td>\n",
       "      <td>psbs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13</td>\n",
       "      <td>14</td>\n",
       "      <td>AAGGAACTATGGCATCGAGC</td>\n",
       "      <td>AAGGACTGCATACCAGGTTG</td>\n",
       "      <td>psbs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>15</td>\n",
       "      <td>16</td>\n",
       "      <td>AAGGATATGTAGACACCGCC</td>\n",
       "      <td>AAGGCCCAGAAGGATACAAC</td>\n",
       "      <td>psbs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>17</td>\n",
       "      <td>18</td>\n",
       "      <td>AAGGCGCTCGGATAATACTC</td>\n",
       "      <td>AAGGTATGTATAGCGACCGC</td>\n",
       "      <td>psbs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>20</td>\n",
       "      <td>21</td>\n",
       "      <td>AATAGGAACCTCTTACGCGG</td>\n",
       "      <td>AATATCACGCAAAAGCACCG</td>\n",
       "      <td>raf1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>22</td>\n",
       "      <td>23</td>\n",
       "      <td>AATCAGTTTCTTTGGCAGCC</td>\n",
       "      <td>AATGCAAAGCTATTAGCGCG</td>\n",
       "      <td>raf1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>24</td>\n",
       "      <td>25</td>\n",
       "      <td>AATGCGTCATTTTACACGGC</td>\n",
       "      <td>AATGTCCTTAGGCAGTCGTC</td>\n",
       "      <td>raf1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>26</td>\n",
       "      <td>27</td>\n",
       "      <td>ACAACGAGCAGACCGAATAG</td>\n",
       "      <td>ACAAGGAGTCGGCATATCAC</td>\n",
       "      <td>raf1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>28</td>\n",
       "      <td>29</td>\n",
       "      <td>ACAGAACGAACAGGCACTAC</td>\n",
       "      <td>ACAGGAAGCAAGGTATACGC</td>\n",
       "      <td>raf1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>30</td>\n",
       "      <td>31</td>\n",
       "      <td>ACAGGGTATATTGAGTGCCC</td>\n",
       "      <td>ACATAAGCGATCCCAAGGTC</td>\n",
       "      <td>raf1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>32</td>\n",
       "      <td>33</td>\n",
       "      <td>ACATCGCATACCAGAACAGG</td>\n",
       "      <td>ACATTAAATTTCGCCGTGGC</td>\n",
       "      <td>raf1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>34</td>\n",
       "      <td>35</td>\n",
       "      <td>ACCACAGGTCAAGATTCACG</td>\n",
       "      <td>ACCCGTATCGCATAAGGATG</td>\n",
       "      <td>zmm28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>36</td>\n",
       "      <td>37</td>\n",
       "      <td>ACGAGATGATGCACCGATAG</td>\n",
       "      <td>ACGATGGGGACATAGAACAC</td>\n",
       "      <td>zmm28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>38</td>\n",
       "      <td>39</td>\n",
       "      <td>ACGGAGCCCTTATTGTAACC</td>\n",
       "      <td>ACGTATGGGGAACACTACAC</td>\n",
       "      <td>zmm28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>40</td>\n",
       "      <td>41</td>\n",
       "      <td>ACGTGAAACTGTATCGAGCC</td>\n",
       "      <td>ACGTTCAGTTTTCCAATGGC</td>\n",
       "      <td>zmm28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>42</td>\n",
       "      <td>43</td>\n",
       "      <td>ACTAGATTAGCAAGGCACCC</td>\n",
       "      <td>ACTGGACCCAATAAAAGGCC</td>\n",
       "      <td>zmm28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>44</td>\n",
       "      <td>45</td>\n",
       "      <td>ACTTCGATTGGCAAGGACTG</td>\n",
       "      <td>AGAACATAGCATTCACGGGG</td>\n",
       "      <td>zmm28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>46</td>\n",
       "      <td>47</td>\n",
       "      <td>AGACAACAATCTGAGGCTGG</td>\n",
       "      <td>AGACAAGCCTTAACCGTAGG</td>\n",
       "      <td>zmm28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>48</td>\n",
       "      <td>49</td>\n",
       "      <td>AGACACAAGGCTGATTCCAG</td>\n",
       "      <td>AGACATGGGATTGACCACAC</td>\n",
       "      <td>sbpase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>50</td>\n",
       "      <td>51</td>\n",
       "      <td>AGAGAGGCATGATTGACCTC</td>\n",
       "      <td>AGAGTTGCACCTAGAATCCG</td>\n",
       "      <td>sbpase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>52</td>\n",
       "      <td>53</td>\n",
       "      <td>AGATAGATGCTCCGTCAAGC</td>\n",
       "      <td>AGATAGTCACGCACAAGACC</td>\n",
       "      <td>sbpase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>54</td>\n",
       "      <td>55</td>\n",
       "      <td>AGATTAGCCGACTTTCCTGG</td>\n",
       "      <td>AGATTAGCTGCCGATACTGG</td>\n",
       "      <td>sbpase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>56</td>\n",
       "      <td>57</td>\n",
       "      <td>AGATTGTTACTCCGACGGAC</td>\n",
       "      <td>AGATTTCCGACGAGATTCCC</td>\n",
       "      <td>sbpase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>58</td>\n",
       "      <td>59</td>\n",
       "      <td>AGCATCCGTCTAAATCTCGG</td>\n",
       "      <td>AGCTATAAGAATTGCCGGGC</td>\n",
       "      <td>sbpase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>60</td>\n",
       "      <td>61</td>\n",
       "      <td>AGCTATGATCCCGGTGTAAC</td>\n",
       "      <td>AGCTCAATCTAACAGTGGGG</td>\n",
       "      <td>sbpase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>62</td>\n",
       "      <td>63</td>\n",
       "      <td>AGGACACCAGACCAATGAAG</td>\n",
       "      <td>AGGGCTAATTACCATCAGCG</td>\n",
       "      <td>sbpase</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   p1  p2                p1_seq                p2_seq    gene\n",
       "0   1   2  AAACACGTGGCAAACATTCC  AAACCGGAGCCATACAGTAC    psbs\n",
       "1   3   4  AAAGCACTCTTAGGCCTCTG  AAAGGGGCCGTCAATATCAG    psbs\n",
       "2   5   6  AAATAAGACGACGACCCTCG  AACGATGATGCTCACTCTCG    psbs\n",
       "3  11  12  AAGAATTACTGACCCCTCGG  AAGACGATCCGAGCCATTAC    psbs\n",
       "4  13  14  AAGGAACTATGGCATCGAGC  AAGGACTGCATACCAGGTTG    psbs\n",
       "5  15  16  AAGGATATGTAGACACCGCC  AAGGCCCAGAAGGATACAAC    psbs\n",
       "6  17  18  AAGGCGCTCGGATAATACTC  AAGGTATGTATAGCGACCGC    psbs\n",
       "0  20  21  AATAGGAACCTCTTACGCGG  AATATCACGCAAAAGCACCG    raf1\n",
       "1  22  23  AATCAGTTTCTTTGGCAGCC  AATGCAAAGCTATTAGCGCG    raf1\n",
       "2  24  25  AATGCGTCATTTTACACGGC  AATGTCCTTAGGCAGTCGTC    raf1\n",
       "3  26  27  ACAACGAGCAGACCGAATAG  ACAAGGAGTCGGCATATCAC    raf1\n",
       "4  28  29  ACAGAACGAACAGGCACTAC  ACAGGAAGCAAGGTATACGC    raf1\n",
       "5  30  31  ACAGGGTATATTGAGTGCCC  ACATAAGCGATCCCAAGGTC    raf1\n",
       "6  32  33  ACATCGCATACCAGAACAGG  ACATTAAATTTCGCCGTGGC    raf1\n",
       "0  34  35  ACCACAGGTCAAGATTCACG  ACCCGTATCGCATAAGGATG   zmm28\n",
       "1  36  37  ACGAGATGATGCACCGATAG  ACGATGGGGACATAGAACAC   zmm28\n",
       "2  38  39  ACGGAGCCCTTATTGTAACC  ACGTATGGGGAACACTACAC   zmm28\n",
       "3  40  41  ACGTGAAACTGTATCGAGCC  ACGTTCAGTTTTCCAATGGC   zmm28\n",
       "4  42  43  ACTAGATTAGCAAGGCACCC  ACTGGACCCAATAAAAGGCC   zmm28\n",
       "5  44  45  ACTTCGATTGGCAAGGACTG  AGAACATAGCATTCACGGGG   zmm28\n",
       "6  46  47  AGACAACAATCTGAGGCTGG  AGACAAGCCTTAACCGTAGG   zmm28\n",
       "0  48  49  AGACACAAGGCTGATTCCAG  AGACATGGGATTGACCACAC  sbpase\n",
       "1  50  51  AGAGAGGCATGATTGACCTC  AGAGTTGCACCTAGAATCCG  sbpase\n",
       "2  52  53  AGATAGATGCTCCGTCAAGC  AGATAGTCACGCACAAGACC  sbpase\n",
       "3  54  55  AGATTAGCCGACTTTCCTGG  AGATTAGCTGCCGATACTGG  sbpase\n",
       "4  56  57  AGATTGTTACTCCGACGGAC  AGATTTCCGACGAGATTCCC  sbpase\n",
       "5  58  59  AGCATCCGTCTAAATCTCGG  AGCTATAAGAATTGCCGGGC  sbpase\n",
       "6  60  61  AGCTATGATCCCGGTGTAAC  AGCTCAATCTAACAGTGGGG  sbpase\n",
       "7  62  63  AGGACACCAGACCAATGAAG  AGGGCTAATTACCATCAGCG  sbpase"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_primers_all = pd.DataFrame()\n",
    "\n",
    "df_primers_psbs = get_primers_df(df_filtered_psbs, 'psbs')\n",
    "df_primers_all = pd.concat([df_primers_all,df_primers_psbs],axis=0)\n",
    "df_primers_raf1 = get_primers_df(df_filtered_raf1, 'raf1')\n",
    "df_primers_all = pd.concat([df_primers_all,df_primers_raf1],axis=0)\n",
    "df_primers_zmm28 = get_primers_df(df_filtered_zmm28, 'zmm28')\n",
    "df_primers_all = pd.concat([df_primers_all,df_primers_zmm28],axis=0)\n",
    "df_primers_sbpase = get_primers_df(df_filtered_sbpase, 'sbpase')\n",
    "df_primers_all = pd.concat([df_primers_all,df_primers_sbpase],axis=0)\n",
    "\n",
    "\n",
    "df_primers_all"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## get backbone vector primers to insert libraries at"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "psbs curr_primers 1 2\n",
      "-1288 GTCCGGAACC\n",
      "-1288 CAAACATTGT\n",
      "1\n",
      "psbs curr_primers 3 4\n",
      "-1088 CAAACATTGT\n",
      "-1088 CTTGGTGAAG\n",
      "2\n",
      "psbs curr_primers 5 6\n",
      "-888 CTTGGTGAAG\n",
      "-888 ACTTCATTCA\n",
      "3\n",
      "psbs curr_primers 11 12\n",
      "-688 ACTTCATTCA\n",
      "-688 CGGCGCGGCC\n",
      "4\n",
      "psbs curr_primers 13 14\n",
      "-488 CGGCGCGGCC\n",
      "-488 AGACACGGGA\n",
      "5\n",
      "doing last segment\n",
      "rev_seq_indices -320 -390\n",
      "GAAAGGGGGCATTGGGGTCACAGGCGCGGGCAGCGTGAAGCTGGAGTGGCGCCAACTGGAGGTGTCGATT\n",
      "psbs curr_primers 15 16\n",
      "-288 AGACACGGGA\n",
      "-288 CGTGAGTAAA\n",
      "6\n",
      "raf1 curr_primers 20 21\n",
      "-1170 AAAAGTTAGA\n",
      "-1170 GGTGGTTATG\n",
      "8\n",
      "raf1 curr_primers 22 23\n",
      "-970 GGTGGTTATG\n",
      "-970 AAGTTGTAAG\n",
      "9\n",
      "raf1 curr_primers 24 25\n",
      "-770 AAGTTGTAAG\n",
      "-770 AACGATCGTA\n",
      "10\n",
      "raf1 curr_primers 26 27\n",
      "-570 AACGATCGTA\n",
      "-570 ATATAGGAAA\n",
      "11\n",
      "raf1 curr_primers 28 29\n",
      "-370 ATATAGGAAA\n",
      "-370 AGACTCTCAT\n",
      "12\n",
      "doing last segment\n",
      "rev_seq_indices -320 -390\n",
      "TCTCCACTCAACGGACGCTTCTCTTCTTCGAGGTTTGAATTTTCCTATATAAAAATTCAGATACGCAACT\n",
      "raf1 curr_primers 30 31\n",
      "-170 AGACTCTCAT\n",
      "-170 AAAGGTTAAC\n",
      "13\n",
      "zmm28 curr_primers 34 35\n",
      "-1247 TCATCAGTCA\n",
      "-1247 GGCAGTGCTC\n",
      "15\n",
      "zmm28 curr_primers 36 37\n",
      "-1047 GGCAGTGCTC\n",
      "-1047 ATATTTTGAT\n",
      "16\n",
      "zmm28 curr_primers 38 39\n",
      "-847 ATATTTTGAT\n",
      "-847 AGTCCCCCTG\n",
      "17\n",
      "zmm28 curr_primers 40 41\n",
      "-647 AGTCCCCCTG\n",
      "-647 CATCCTTCTG\n",
      "18\n",
      "zmm28 curr_primers 42 43\n",
      "-447 CATCCTTCTG\n",
      "-447 GCACCACTCG\n",
      "19\n",
      "doing last segment\n",
      "rev_seq_indices -320 -390\n",
      "GCGAAACCACGGTGGTTTTGCCTGACCCGCTGTCGTGTTTTGATTGGACGATGAGGTTTCGGAGCGGCGT\n",
      "zmm28 curr_primers 44 45\n",
      "-247 GCACCACTCG\n",
      "-247 TTCTTGTTGA\n",
      "20\n",
      "sbpase curr_primers 48 49\n",
      "-1377 TTCTGCGAGG\n",
      "-1377 TTTGGTTTGT\n",
      "22\n",
      "sbpase curr_primers 50 51\n",
      "-1177 TTTGGTTTGT\n",
      "-1177 AATGTGACAC\n",
      "23\n",
      "sbpase curr_primers 52 53\n",
      "-977 AATGTGACAC\n",
      "-977 TAGAAAAGGA\n",
      "24\n",
      "sbpase curr_primers 54 55\n",
      "-777 TAGAAAAGGA\n",
      "-777 TTAAATACCA\n",
      "25\n",
      "sbpase curr_primers 56 57\n",
      "-577 TTAAATACCA\n",
      "-577 GTCATCGTGC\n",
      "26\n",
      "sbpase curr_primers 58 59\n",
      "-377 GTCATCGTGC\n",
      "-377 ACAGCATCCA\n",
      "27\n",
      "doing last segment\n",
      "rev_seq_indices -320 -390\n",
      "GCCATTTGGAGCTGGAGAAGATAGAAAGGGGAGGGCCGCGCTGCGGTGCACGATGACCATTCGTGGTTTG\n",
      "sbpase curr_primers 60 61\n",
      "-177 ACAGCATCCA\n",
      "-177 GAAATTATTT\n",
      "28\n"
     ]
    }
   ],
   "source": [
    "# get backbone sequences\n",
    "c = 0\n",
    "\n",
    "###\n",
    "c=0\n",
    "d_bb_seq_psbs, c = get_matching_backbone_seqs_library(\n",
    "    dic_dic_gene_utr['psbs']['prime_5'], \n",
    "    dic_dic_gene_utr['psbs']['utr_5'], \n",
    "    dic_dic_gene_utr['psbs']['first_10_aa'], \n",
    "    c, \n",
    "    'psbs'\n",
    ")\n",
    "\n",
    "d_bb_seq_raf1, c = get_matching_backbone_seqs_library(\n",
    "    dic_dic_gene_utr['raf1']['prime_5'], \n",
    "    dic_dic_gene_utr['raf1']['utr_5'], \n",
    "    dic_dic_gene_utr['raf1']['first_10_aa'], \n",
    "    c, 'raf1')\n",
    "\n",
    "d_bb_seq_zmm28, c = get_matching_backbone_seqs_library(\n",
    "    dic_dic_gene_utr['zmm28']['prime_5'], \n",
    "    dic_dic_gene_utr['zmm28']['utr_5'], \n",
    "    dic_dic_gene_utr['zmm28']['first_10_aa'], \n",
    "    c, 'zmm28')\n",
    "\n",
    "d_bb_seq_sbpase, c = get_matching_backbone_seqs_library(\n",
    "    dic_dic_gene_utr['sbpase']['prime_5'], \n",
    "    dic_dic_gene_utr['sbpase']['utr_5'], \n",
    "    dic_dic_gene_utr['sbpase']['first_10_aa'], \n",
    "    c, 'sbpase')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1_2_rev 63.087897914902555\n",
      "1_2_fwd 63.018726488303344\n",
      "3_4_rev 63.02025148890124\n",
      "3_4_fwd 63.467712419348175\n",
      "5_6_rev 63.46038869723145\n",
      "5_6_fwd 63.76186304611872\n",
      "11_12_rev 63.05896958824479\n",
      "11_12_fwd 63.7482917332502\n",
      "13_14_rev 64.17936197199913\n",
      "13_14_fwd 64.63413487862323\n",
      "15_16_rev 63.64825209752297\n",
      "15_16_fwd 63.37195169691586\n",
      "17_18_rev 63.18999273666401\n",
      "17_18_fwd 63.582471892771764\n",
      "hit total length\n",
      "20_21_rev 60.117187786263614\n",
      "20_21_fwd 63.30455099155819\n",
      "hit total length\n",
      "22_23_rev 59.53147350054934\n",
      "22_23_fwd 63.13819618962496\n",
      "hit total length\n",
      "24_25_rev 58.94575921483505\n",
      "24_25_fwd 63.498914156880346\n",
      "26_27_rev 63.07030072567778\n",
      "26_27_fwd 63.34146386778747\n",
      "28_29_rev 63.015222618389885\n",
      "28_29_fwd 63.26979030259878\n",
      "30_31_rev 63.23553213889329\n",
      "30_31_fwd 63.50430935335055\n",
      "32_33_rev 63.23553213889329\n",
      "32_33_fwd 63.50259460612091\n",
      "34_35_rev 63.68376149489325\n",
      "34_35_fwd 63.43838197116378\n",
      "36_37_rev 63.170871066519226\n",
      "hit total length\n",
      "36_37_fwd 62.46004492912077\n",
      "38_39_rev 63.32604414810828\n",
      "38_39_fwd 63.382041257300386\n",
      "40_41_rev 63.20716977377623\n",
      "40_41_fwd 64.3649489802591\n",
      "42_43_rev 63.21792422103181\n",
      "42_43_fwd 63.70645442844017\n",
      "44_45_rev 63.12994433381368\n",
      "44_45_fwd 64.73010432826118\n",
      "46_47_rev 63.973335720273724\n",
      "46_47_fwd 63.80481520764579\n",
      "48_49_rev 63.423447280048265\n",
      "48_49_fwd 63.01837436714658\n",
      "50_51_rev 63.17540740552249\n",
      "50_51_fwd 63.294326472046464\n",
      "hit total length\n",
      "52_53_rev 60.117187786263614\n",
      "52_53_fwd 63.75752593862205\n",
      "54_55_rev 63.224323125881824\n",
      "54_55_fwd 63.31429851143912\n",
      "56_57_rev 63.36578063494261\n",
      "56_57_fwd 63.98730604148085\n",
      "58_59_rev 63.24087071301955\n",
      "58_59_fwd 64.07881509559195\n",
      "60_61_rev 63.12421103028487\n",
      "60_61_fwd 63.1073349736572\n",
      "62_63_rev 63.121656268681875\n",
      "62_63_fwd 64.74369689106953\n"
     ]
    }
   ],
   "source": [
    "d_bb_seq_psbs_keep = get_tm_bb_primers(d_bb_seq_psbs)\n",
    "d_bb_seq_raf1_keep = get_tm_bb_primers(d_bb_seq_raf1)\n",
    "d_bb_seq_sbpase_keep = get_tm_bb_primers(d_bb_seq_sbpase)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add aarI for PsbS\n",
    "d_bb_seq_psbs_final = add_t2s_sites(d_bb_seq_psbs_keep, 'aar1')\n",
    "df_bb_prims_all = pd.DataFrame.from_dict(d_bb_seq_psbs_final, orient='index')\n",
    "df_bb_prims_all['gene'] = 'psbs'\n",
    "\n",
    "# add bsaI for Raf1\n",
    "d_bb_seq_raf1_final = add_t2s_sites(d_bb_seq_raf1_keep, 'bsa1')\n",
    "df_bb_prims = pd.DataFrame.from_dict(d_bb_seq_raf1_final, orient='index')\n",
    "df_bb_prims['gene'] = 'raf1'\n",
    "df_bb_prims_all = pd.concat([df_bb_prims_all, df_bb_prims], axis=0)\n",
    "\n",
    "# add bsmbI for SBPase\n",
    "d_bb_seq_sbpase_final = add_t2s_sites(d_bb_seq_sbpase_keep, 'bsmb1')\n",
    "df_bb_prims = pd.DataFrame.from_dict(d_bb_seq_sbpase_final, orient='index')\n",
    "df_bb_prims['gene'] = 'sbpase'\n",
    "df_bb_prims_all = pd.concat([df_bb_prims_all, df_bb_prims], axis=0)\n",
    "\n",
    "\n",
    "# add an ID, and write to file\n",
    "df_bb_prims_all['primer_n'] = df_bb_prims_all.index + '_'+df_bb_prims_all.gene +'_bb'\n",
    "df_bb_prims_all = df_bb_prims_all.rename(columns={'0':'seq'})\n",
    "\n",
    "df_bb_prims_fwd = df_bb_prims_all.loc[df_bb_prims_all.index.str.endswith('fwd')]\n",
    "df_bb_prims_rev = df_bb_prims_all.loc[df_bb_prims_all.index.str.endswith('rev')]\n",
    "\n",
    "df_bb_prims_fwd.to_csv('df_bb_prims_fwd.csv')\n",
    "df_bb_prims_rev.to_csv('df_bb_prims_rev.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# get primers for amplification of ordered sublibraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/davidding/anaconda3/envs/tf_env/lib/python3.7/site-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  after removing the cwd from sys.path.\n",
      "/Users/davidding/anaconda3/envs/tf_env/lib/python3.7/site-packages/ipykernel_launcher.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  \"\"\"\n"
     ]
    }
   ],
   "source": [
    "df_primers_all['primers'] = df_primers_all['p1'].astype(str) + '_' + df_primers_all['p2'].astype(str)\n",
    "df_primers_chunks_fwd = df_primers_all[['p1_seq', 'gene', 'primers']]\n",
    "df_primers_chunks_rev = df_primers_all[['p2_seq', 'gene', 'primers']]\n",
    "df_primers_chunks_rev['primer_name']= df_primers_chunks_rev['primers'] +df_primers_chunks_rev['gene'] +'_rev_chunk'\n",
    "df_primers_chunks_fwd['primer_name']= df_primers_chunks_fwd['primers'] + df_primers_chunks_rev['gene']+'_fwd_chunk'\n",
    "df_primers_chunks_fwd.to_csv('df_primers_chunks_fwd.csv')\n",
    "df_primers_chunks_rev.to_csv('df_primers_chunks_rev.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# create primers to make around the horn large deletions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "psbs [('aar1', 'CACCTGC')]\n",
      "198\n",
      "raf1 [('bsa1', 'GGTCTC'), ('aar1', 'CACCTGC')]\n",
      "1383\n",
      "zmm28 [('bsa1', 'GGTCTC')]\n",
      "185\n",
      "sbpase [('bsmb1', 'CGTCTC')]\n",
      "75\n",
      "psbs [('aar1', 'CACCTGC')]\n",
      "198\n",
      "raf1 [('bsa1', 'GGTCTC'), ('aar1', 'CACCTGC')]\n",
      "1383\n",
      "zmm28 [('bsa1', 'GGTCTC')]\n",
      "185\n",
      "sbpase [('bsmb1', 'CGTCTC')]\n",
      "75\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'g_psbs_p_-2198_dsize_200_re_aar1_f': 'gatacaCACCTGCaccaacacatgctaatgctaagatctctac',\n",
       " 'g_psbs_p_-2198_dsize_200_re_aar1_r': 'GATACAcacctgcGCATGTGTctagaaatatgttgagcagattcattcctgcc',\n",
       " 'g_psbs_p_-1998_dsize_200_re_aar1_f': 'gatacaCACCTGCgaggctacaatgctcttatttatttaaatatttgatttgc',\n",
       " 'g_psbs_p_-1998_dsize_200_re_aar1_r': 'GATACAcacctgcCATTGTAGtggtatacataaagaatgcatacataaatagcctctagt',\n",
       " 'g_psbs_p_-1798_dsize_200_re_aar1_f': 'gatacaCACCTGCggaaagcaagcatttgtatatggatcttatgc',\n",
       " 'g_psbs_p_-1798_dsize_200_re_aar1_r': 'GATACAcacctgcTGCTTGCTcctcacagggccaatccacaagttgcttcg',\n",
       " 'g_psbs_p_-1598_dsize_200_re_aar1_f': 'gatacaCACCTGCtctatcatgaacaaccagcacatggaccgt',\n",
       " 'g_psbs_p_-1598_dsize_200_re_aar1_r': 'GATACAcacctgcGTTCATGAttccgttgtgggtggacccagaacacgtca',\n",
       " 'g_psbs_p_-1398_dsize_200_re_aar1_f': 'gatacaCACCTGCgtttctgcctactgtagcactttcgttttt',\n",
       " 'g_psbs_p_-1398_dsize_200_re_aar1_r': 'GATACAcacctgcGTAGGCAGtagagataaatatcaccatcagcacatcagtct',\n",
       " 'g_psbs_p_-1198_dsize_200_re_aar1_f': 'gatacaCACCTGCaggccttacagcaacattgccacagcgggt',\n",
       " 'g_psbs_p_-1198_dsize_200_re_aar1_r': 'GATACAcacctgcGCTGTAAGaaaccaaagtttttgccaactaaacaaggc',\n",
       " 'g_psbs_p_-998_dsize_200_re_aar1_f': 'gatacaCACCTGCaccgtcactaacagacagtagaatggactc',\n",
       " 'g_psbs_p_-998_dsize_200_re_aar1_r': 'GATACAcacctgcGTTAGTGAgccttgtttagttcgcaaaaattttcaagatt',\n",
       " 'g_psbs_p_-798_dsize_200_re_aar1_f': 'gatacaCACCTGCcacgcacgctggctcctgggctccctggcc',\n",
       " 'g_psbs_p_-798_dsize_200_re_aar1_r': 'GATACAcacctgcCCAGCGTGcggttgtcatgtatggagcagatcgtagca',\n",
       " 'g_psbs_p_-598_dsize_200_re_aar1_f': 'gatacaCACCTGCtttcataagaacgggacacgcggcctcgga',\n",
       " 'g_psbs_p_-598_dsize_200_re_aar1_r': 'GATACAcacctgcGTTCTTATcgtgcggctcctgtcctgcgacctccgtac',\n",
       " 'g_psbs_p_-398_dsize_200_re_aar1_f': 'gatacaCACCTGCacatatggcgcagtccatgctcatgtcgac',\n",
       " 'g_psbs_p_-398_dsize_200_re_aar1_r': 'GATACAcacctgcTGCGCCATgaaagggggcattggggtcacaggcgcggg',\n",
       " 'g_raf1_p_-3383_dsize_200_re_bsa1_f': 'gatacaGGTCTCattgaaaaccatacaaattatatagctttgtacacatttagc',\n",
       " 'g_raf1_p_-3383_dsize_200_re_bsa1_r': 'GATACAggtctcTTCAATtgaaaacttgaaatatgtggtacactggtact',\n",
       " 'g_raf1_p_-3183_dsize_200_re_bsa1_f': 'gatacaGGTCTCactcttaatatctactttcagattacatataaaaaacgattgga',\n",
       " 'g_raf1_p_-3183_dsize_200_re_bsa1_r': 'GATACAggtctcAAGAGTttcggatataaatgtttgcaataagtaactagaaacttaga',\n",
       " 'g_raf1_p_-2983_dsize_200_re_bsa1_f': 'gatacaGGTCTCttttcttttttgaaaaattataatcacaacttttacgtataacttatgg',\n",
       " 'g_raf1_p_-2983_dsize_200_re_bsa1_r': 'GATACAggtctcAGAAAActttttcgttccccttgctaaatttagtcag',\n",
       " 'g_raf1_p_-2783_dsize_200_re_bsa1_f': 'gatacaGGTCTCtatcaacataatttatgacgagacaatttagcgaac',\n",
       " 'g_raf1_p_-2783_dsize_200_re_bsa1_r': 'GATACAggtctcTTGATAgtagagattttttttgcctaaaatagaaaactttccaaaa',\n",
       " 'g_raf1_p_-2583_dsize_200_re_bsa1_f': 'gatacaGGTCTCacacttacacgtagaaagttagatataaaagttgcg',\n",
       " 'g_raf1_p_-2583_dsize_200_re_bsa1_r': 'GATACAggtctcAAGTGTaacttgttccataaaatcttatattatagataaaagttatgcgtataac',\n",
       " 'g_raf1_p_-2383_dsize_200_re_bsa1_f': 'gatacaGGTCTCaatcgtgaacgatcacccaaataaaataaggg',\n",
       " 'g_raf1_p_-2383_dsize_200_re_bsa1_r': 'GATACAggtctcACGATTgtaactagtagatattaacttataattattatcttctttgtctaactttttgc',\n",
       " 'g_raf1_p_-2183_dsize_200_re_bsa1_f': 'gatacaGGTCTCgttataccaagttcttagtataatttatagaaataagttatgttctataacatatatgt',\n",
       " 'g_raf1_p_-2183_dsize_200_re_bsa1_r': 'GATACAggtctcTATAACtgatgtaacaaaggttcaaactacctctcc',\n",
       " 'g_raf1_p_-1983_dsize_200_re_bsa1_f': 'gatacaGGTCTCaacatgatgcttggactgtgttatgctttt',\n",
       " 'g_raf1_p_-1983_dsize_200_re_bsa1_r': 'GATACAggtctcCATGTTactttgcaaaacttatcaccatatgacttatacca',\n",
       " 'g_raf1_p_-1783_dsize_200_re_bsa1_f': 'gatacaGGTCTCgagaaggctcgattcagaagcaagcgcgac',\n",
       " 'g_raf1_p_-1783_dsize_200_re_bsa1_r': 'GATACAggtctcCTTCTCtaatagtcgcaaggcaaaatcatagacaca',\n",
       " 'g_raf1_p_-1583_dsize_200_re_bsa1_f': 'gatacaGGTCTCatgctgtcactctcgcacccgcaccccgcc',\n",
       " 'g_raf1_p_-1583_dsize_200_re_bsa1_r': 'GATACAggtctcCAGCATtctccactcaacggacgcttctcttcttcg',\n",
       " 'g_zmm28_p_-2185_dsize_200_re_bsa1_f': 'gatacaGGTCTCctgagatcagctggtgcatcctggtttaag',\n",
       " 'g_zmm28_p_-2185_dsize_200_re_bsa1_r': 'GATACAggtctcTCTCAGactgctccacgaactctgctccataaactc',\n",
       " 'g_zmm28_p_-1985_dsize_200_re_bsa1_f': 'gatacaGGTCTCcataggaataagatgtgcacgtgtttataaagatg',\n",
       " 'g_zmm28_p_-1985_dsize_200_re_bsa1_r': 'GATACAggtctcCCTATGctacatcggttcgtctaataaacacatttatttgac',\n",
       " 'g_zmm28_p_-1785_dsize_200_re_bsa1_f': 'gatacaGGTCTCaatgttataattttctacccatccataaaatcatgatttcc',\n",
       " 'g_zmm28_p_-1785_dsize_200_re_bsa1_r': 'GATACAggtctcAACATTaggatattttgaaggactggtactaaccagg',\n",
       " 'g_zmm28_p_-1585_dsize_200_re_bsa1_f': 'gatacaGGTCTCatactcctagctcactcttgagcatcacca',\n",
       " 'g_zmm28_p_-1585_dsize_200_re_bsa1_r': 'GATACAggtctcGAGTATtaaaggatacttatgggcaaagacaaaattaactgt',\n",
       " 'g_zmm28_p_-1385_dsize_200_re_bsa1_f': 'gatacaGGTCTCtagcgccactaggttgatgaggtgcagaaa',\n",
       " 'g_zmm28_p_-1385_dsize_200_re_bsa1_r': 'GATACAggtctcGCGCTAgggcggtggtggagaatgacaatgggagca',\n",
       " 'g_zmm28_p_-1185_dsize_200_re_bsa1_f': 'gatacaGGTCTCgtttacattttagctatttttctagctcttgctacaa',\n",
       " 'g_zmm28_p_-1185_dsize_200_re_bsa1_r': 'GATACAggtctcGTAAACgcaagactagaggacctcaccgtgatggtt',\n",
       " 'g_zmm28_p_-985_dsize_200_re_bsa1_f': 'gatacaGGTCTCagtccagaaatggaagaaaaagggcagagc',\n",
       " 'g_zmm28_p_-985_dsize_200_re_bsa1_r': 'GATACAggtctcTGGACTctagtagtttatttctccaagagagcaaataaaaacatg',\n",
       " 'g_zmm28_p_-785_dsize_200_re_bsa1_f': 'gatacaGGTCTCacgacgggctcgaccaatccctgccccgcg',\n",
       " 'g_zmm28_p_-785_dsize_200_re_bsa1_r': 'GATACAggtctcCGTCGTgcaacatccccaaattggttgccgaaattg',\n",
       " 'g_zmm28_p_-585_dsize_200_re_bsa1_f': 'gatacaGGTCTCgataccgctctcctccccatctaaaaccgc',\n",
       " 'g_zmm28_p_-585_dsize_200_re_bsa1_r': 'GATACAggtctcGGTATCgggccggaaagggtgcggctgtgggtagcg',\n",
       " 'g_zmm28_p_-385_dsize_200_re_bsa1_f': 'gatacaGGTCTCatggggcgcggaccggtgcagctgcgccgg',\n",
       " 'g_zmm28_p_-385_dsize_200_re_bsa1_r': 'GATACAggtctcCCCCATgcgaaaccacggtggttttgcctgacccgc',\n",
       " 'g_sbpase_p_-2075_dsize_200_re_bsmb1_f': 'gatacaCGTCTCggctcatggacatgtactgcgcggcaagca',\n",
       " 'g_sbpase_p_-2075_dsize_200_re_bsmb1_r': 'GATACAcgtctcTGAGCCctggccggcgacggcgtcgacattgatcct',\n",
       " 'g_sbpase_p_-1875_dsize_200_re_bsmb1_f': 'gatacaCGTCTCgccgaggctttttcgggtttgatccggcac',\n",
       " 'g_sbpase_p_-1875_dsize_200_re_bsmb1_r': 'GATACAcgtctcCTCGGCgccgcagcttcaagtcgcgcttgacgcgga',\n",
       " 'g_sbpase_p_-1675_dsize_200_re_bsmb1_f': 'gatacaCGTCTCtgatatattgtgatactagttcaatcatagaggcct',\n",
       " 'g_sbpase_p_-1675_dsize_200_re_bsmb1_r': 'GATACAcgtctcATATCAtagtcggctctcacgcagaagtatgcctcg',\n",
       " 'g_sbpase_p_-1475_dsize_200_re_bsmb1_f': 'gatacaCGTCTCcattgcaattaggtagaagagtttaaccatacaga',\n",
       " 'g_sbpase_p_-1475_dsize_200_re_bsmb1_r': 'GATACAcgtctcGCAATGagtgaaacagatcagcgaagtcacgatctc',\n",
       " 'g_sbpase_p_-1275_dsize_200_re_bsmb1_f': 'gatacaCGTCTCtctatttagataatatatgcattagatattatgatagtatttctcataataaatttaaaaatataaattttataac',\n",
       " 'g_sbpase_p_-1275_dsize_200_re_bsmb1_r': 'GATACAcgtctcAATAGAatatacaaatattttatgtatttgtgaaaaaaaatgtttgacttggg',\n",
       " 'g_sbpase_p_-1075_dsize_200_re_bsmb1_f': 'gatacaCGTCTCctgctcatgttatgactcaaagttgctaaag',\n",
       " 'g_sbpase_p_-1075_dsize_200_re_bsmb1_r': 'GATACAcgtctcGAGCAGttattaattataaatggttggtcaagtatttcaatgtttgact',\n",
       " 'g_sbpase_p_-875_dsize_200_re_bsmb1_f': 'gatacaCGTCTCtaaatactagaaccaaacatcccagaccga',\n",
       " 'g_sbpase_p_-875_dsize_200_re_bsmb1_r': 'GATACAcgtctcTATTTAgagttcaagcctgcccagacagaattactc',\n",
       " 'g_sbpase_p_-675_dsize_200_re_bsmb1_f': 'gatacaCGTCTCttaaataccagaacagagccaaacaaacatcc',\n",
       " 'g_sbpase_p_-675_dsize_200_re_bsmb1_r': 'GATACAcgtctcATTTAAattgtcgtatatatacgagtacgacggtttcg',\n",
       " 'g_sbpase_p_-475_dsize_200_re_bsmb1_f': 'gatacaCGTCTCgccaccgcggcggccgggttatcttgtcca',\n",
       " 'g_sbpase_p_-475_dsize_200_re_bsmb1_r': 'GATACAcgtctcGGTGGCttgtcgtatatacgatatggatcattttttctgaaattg',\n",
       " 'g_sbpase_p_-275_dsize_200_re_bsmb1_f': 'gatacaCGTCTCatggagatcgtcgccacgcgctccccggcc',\n",
       " 'g_sbpase_p_-275_dsize_200_re_bsmb1_r': 'GATACAcgtctcCTCCATgccatttggagctggagaagatagaaaggg',\n",
       " 'g_psbs_p_-2198_dsize_500_re_aar1_f': 'gatacaCACCTGCtctactgtattacggttagggaagcaagaa',\n",
       " 'g_psbs_p_-2198_dsize_500_re_aar1_r': 'GATACAcacctgcTAATACAGctagaaatatgttgagcagattcattcctgcc',\n",
       " 'g_psbs_p_-1698_dsize_500_re_aar1_f': 'gatacaCACCTGCgtttctgcctactgtagcactttcgttttt',\n",
       " 'g_psbs_p_-1698_dsize_500_re_aar1_r': 'GATACAcacctgcGTAGGCAGtagagccaatgcatataaacaaaaaaaaacacaca',\n",
       " 'g_psbs_p_-1198_dsize_500_re_aar1_f': 'gatacaCACCTGCatctcaccgaccgaccagcgccatcagttt',\n",
       " 'g_psbs_p_-1198_dsize_500_re_aar1_r': 'GATACAcacctgcGGTCGGTGaaaccaaagtttttgccaactaaacaaggc',\n",
       " 'g_psbs_p_-698_dsize_500_re_aar1_f': 'gatacaCACCTGCacatatggcgcagtccatgctcatgtcgac',\n",
       " 'g_psbs_p_-698_dsize_500_re_aar1_r': 'GATACAcacctgcTGCGCCATagatacacgcgtcaagcaaaaagtgttctg',\n",
       " 'g_raf1_p_-3383_dsize_500_re_bsa1_f': 'gatacaGGTCTCgatcagacaactctcacgactttccttttt',\n",
       " 'g_raf1_p_-3383_dsize_500_re_bsa1_r': 'GATACAggtctcCTGATCtgaaaacttgaaatatgtggtacactggtact',\n",
       " 'g_raf1_p_-2883_dsize_500_re_bsa1_f': 'gatacaGGTCTCacacttacacgtagaaagttagatataaaagttgcg',\n",
       " 'g_raf1_p_-2883_dsize_500_re_bsa1_r': 'GATACAggtctcAAGTGTgttcgctagattcaaatctattgaacgttcg',\n",
       " 'g_raf1_p_-2383_dsize_500_re_bsa1_f': 'gatacaGGTCTCattatttccaaataagaaaagttgcgggagtagc',\n",
       " 'g_raf1_p_-2383_dsize_500_re_bsa1_r': 'GATACAggtctcAATAATgtaactagtagatattaacttataattattatcttctttgtctaactttttgc',\n",
       " 'g_raf1_p_-1883_dsize_500_re_bsa1_f': 'gatacaGGTCTCatgctgtcactctcgcacccgcaccccgcc',\n",
       " 'g_raf1_p_-1883_dsize_500_re_bsa1_r': 'GATACAggtctcCAGCATaatttccgcaacttttcatttttagaaggtaact',\n",
       " 'g_zmm28_p_-2185_dsize_500_re_bsa1_f': 'gatacaGGTCTCatgcgcttagtgcaaatttgttcgtggaca',\n",
       " 'g_zmm28_p_-2185_dsize_500_re_bsa1_r': 'GATACAggtctcGCGCATactgctccacgaactctgctccataaactc',\n",
       " 'g_zmm28_p_-1685_dsize_500_re_bsa1_f': 'gatacaGGTCTCtagcgccactaggttgatgaggtgcagaaa',\n",
       " 'g_zmm28_p_-1685_dsize_500_re_bsa1_r': 'GATACAggtctcGCGCTAttcttcttctttttctttgccaatcatagtataataatacac',\n",
       " 'g_zmm28_p_-1185_dsize_500_re_bsa1_f': 'gatacaGGTCTCacttgccacctcgctcccttgtgcccgaca',\n",
       " 'g_zmm28_p_-1185_dsize_500_re_bsa1_r': 'GATACAggtctcGCAAGTgcaagactagaggacctcaccgtgatggtt',\n",
       " 'g_zmm28_p_-685_dsize_500_re_bsa1_f': 'gatacaGGTCTCatggggcgcggaccggtgcagctgcgccgg',\n",
       " 'g_zmm28_p_-685_dsize_500_re_bsa1_r': 'GATACAggtctcCCCCATgacacgtggggtagcttcagggggactggc',\n",
       " 'g_sbpase_p_-2075_dsize_500_re_bsmb1_f': 'gatacaCGTCTCggaatcctatgcggtgttgattgctcaact',\n",
       " 'g_sbpase_p_-2075_dsize_500_re_bsmb1_r': 'GATACAcgtctcGATTCCctggccggcgacggcgtcgacattgatcct',\n",
       " 'g_sbpase_p_-1575_dsize_500_re_bsmb1_f': 'gatacaCGTCTCtctatttagataatatatgcattagatattatgatagtatttctcataataaatttaaaaatataaattttataac',\n",
       " 'g_sbpase_p_-1575_dsize_500_re_bsmb1_r': 'GATACAcgtctcAATAGAaaaaaagtcatcataatcataagagaagaacatcacac',\n",
       " 'g_sbpase_p_-1075_dsize_500_re_bsmb1_f': 'gatacaCGTCTCccataccgatacggcgataccgtcgtactc',\n",
       " 'g_sbpase_p_-1075_dsize_500_re_bsmb1_r': 'GATACAcgtctcGTATGGttattaattataaatggttggtcaagtatttcaatgtttgact',\n",
       " 'g_sbpase_p_-575_dsize_500_re_bsmb1_f': 'gatacaCGTCTCatggagatcgtcgccacgcgctccccggcc',\n",
       " 'g_sbpase_p_-575_dsize_500_re_bsmb1_r': 'GATACAcgtctcCTCCATaactttttctgaattgttcagtacttcagtgtg'}"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prime5_full = dic_dic_gene_utr['raf1']['prime_5'] + dic_dic_gene_utr['raf1']['utr_5']\n",
    "exon_1 = dic_dic_gene_utr['raf1']['exon_1']\n",
    "\n",
    "prime5_exon1_full = prime5_full + exon_1\n",
    "\n",
    "all_dic = {}\n",
    "# make all 200 bp deletions\n",
    "for gene, gene_dic in dic_dic_gene_utr.items():\n",
    "\n",
    "    d_curr = get_deletions_primers(gene, \n",
    "                                   gene_to_fl_seq[gene],\n",
    "                          gene_dic['prime_5'], \n",
    "                          gene_dic['utr_5'], \n",
    "                          gene_dic['exon_1'], \n",
    "                          chunk_size=200, max_len_primer = 90, min_tm = 57)\n",
    "    all_dic.update(d_curr)\n",
    "\n",
    "# make all 500 bp deletions\n",
    "for gene, gene_dic in dic_dic_gene_utr.items():\n",
    "\n",
    "    d_curr = get_deletions_primers(gene, \n",
    "                                   gene_to_fl_seq[gene],\n",
    "                          gene_dic['prime_5'], \n",
    "                          gene_dic['utr_5'], \n",
    "                          gene_dic['exon_1'], \n",
    "                          chunk_size=500, max_len_primer = 90, min_tm = 57)\n",
    "    all_dic.update(d_curr)\n",
    "\n",
    "df_del = pd.DataFrame.from_dict(all_dic, orient='index')\n",
    "df_del_f = df_del.loc[df_del.index.str[-1]=='f']\n",
    "df_del_r = df_del.loc[df_del.index.str[-1]=='r']\n",
    "df_del_f.to_csv('./del_f_primers_order.csv')\n",
    "df_del_r.to_csv('./del_r_primers_order.csv')\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base2",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
